Server Preparation for Ubuntu 22.04 (#369)
owncloud/docs-server: modules/admin_manual/pages/installation/manual_installation/server_prep_ubuntu_22.04.adoc

= Server Preparation for Ubuntu 22.04
:toc: right
:toclevels: 1
:page-partial:
:ubuntu_upgrade_url: https://linuxconfig.org/how-to-upgrade-ubuntu-to-22-04-lts-jammy-jellyfish
:php-common: https://tecadmin.net/enable-disable-php-modules-ubuntu/
:ondrej-php-url: https://launchpad.net/~ondrej/+archive/ubuntu/php
:libsmbclient-php_url: https://github.com/eduardok/libsmbclient-php
:update-alternatives_url: https://manpages.ubuntu.com/manpages/focal/en/man1/update-alternatives.1.html
:phpmyadmin_latest_url: https://www.phpmyadmin.net/downloads/
:pear-package_url: https://pear.php.net/package/PEAR/

> :page-aliases: installation/manual_installation/server_prep_ubuntu_18.04.adoc

Reviewer: ```suggestion
```
Removing the page alias, as we do not have a legacy here and it will cause build errors (double referencing)

== Introduction

The target audience for this document are experienced admins with additional needs and setup scenarios. Ubuntu 22.04 by default only provides PHP 8.1. However, PHP versions 8.0 and 8.1 are currently not supported, therefore this guide intends to documents the installation of PHP version 7.4 on Ubuntu 22.04 by the means of the {ondrej-php-url}[ondrej/php] PPA. For more information see {php-supported-versions-url}[Currently Supported PHP Versions]. This document considers two basic scenarios. You have/plan to upgrade from 20.04 LTS to 22.04 LTS or you have a fresh Ubuntu 22.04 LTS installation.

> IMPORTANT: PHP 8.0 and 8.1 are not currently supported by the ownCloud server.

Reviewer: ```suggestion
IMPORTANT: PHP 8.x is currently *not* supported by ownCloud server.
```
There may be a 8.2 present in future and we may not support it at all...

[source,bash]
----
sudo apt purge `dpkg -l | grep php | awk '{print $2}' | tr "\n" " "`
----

NOTE: Keep a copy of your relevant `php.ini` and other configuration files as reference _before_ you start purging, to quickly reconfigure your PHP 7.4 installation afterwards.

After deinstallation of all versions and extensions of PHP and cleaning up all remnants, you should have an empty `/etc/php` directory. You can proceed like on a fresh installation of Ubuntu 22.04 LTS.

== Clean Ubuntu 22.04 Installation

After you have installed your Ubuntu 22.04 LTS server from scratch, use the following commands to first add the {ondrej-php-url}[ondrej/php] PPA and then install PHP 7.4 and all necessary extensions:

[source,bash]
----
sudo add-apt-repository ppa:ondrej/php -y
----

[source,bash]
----
sudo apt update && sudo apt upgrade
----

[source,bash]
----
sudo apt install php7.4 php7.4-intl php7.4-mysql php7.4-mbstring \
       php7.4-redis php7.4-imagick php7.4-igbinary php7.4-gmp \
       php7.4-bcmath php7.4-curl php7.4-gd php7.4-zip php7.4-imap \
       php7.4-ldap php7.4-bz2 php7.4-ssh2 php-phpseclib \
       php7.4-common php7.4-json php7.4-xml php7.4-dev php7.4-apcu \

>        php7.4-redis libsmbclient-dev php-pear

Author: To add php7.4-opcache to align with the quick install guide

Reviewer: fixed

>        php7.4-ldap php7.4-bz2 php7.4-ssh2 php-phpseclib \
>        php7.4-common php7.4-json php7.4-xml php7.4-dev php7.4-apcu \
>        php7.4-redis libsmbclient-dev php-pear

Reviewer: ```suggestion
       php7.4-ldap php7.4-bz2 php7.4-ssh2 php7.4-phpseclib \
       php7.4-common php7.4-json php7.4-xml php7.4-dev php7.4-apcu \
       php7.4-redis php7.4-opcache libsmbclient-dev php-pear
```

[source,bash]
----
php -i | grep extension_dir
----

If the output is different than the style below, there is a problem that needs fixing:

[source,plaintext]
----
extension_dir => /usr/lib/php/20180731 => /usr/lib/php/20180731
----

This is the output that shows a fix is needed:

[source,plaintext]
----
PHP Warning:  PHP Startup: smbclient: Unable to initialize module
Module compiled with module API=20190902
PHP    compiled with module API=20180731
These options need to match
----

As you see above, the API modules do not match and have been compiled with different versions and therefore will not work. To fix this, uninstall the PECL module with `pecl uninstall <extension_name>`, set the correct `update-alternatives` as described above and reinstall it.

== libsmbclient-php Library

`libsmbclient-php` is a PHP extension that uses Samba's libsmbclient library to provide Samba-related functions to PHP programs. You need to install it if you have installed `smbclient` as described above. `smbclient` is e.g. necessary if you are using the 
xref:enterprise/external_storage/windows-network-drive_configuration.adoc[Windows Network Drives app]

> from ownCloud's Enterprise Edition. To install it, run the command described below.

Reviewer 2: ```suggestion
from ownCloud's Enterprise Edition. To install it, run the following command:
```

[source,bash]
----
sudo apt install php7.4-smbclient
----

> When the command completes, you then have to:

Reviewer 2: ```suggestion
When the command completes, perform the following tasks:
```

[source,bash]
----
ls `php -i | grep "^extension_dir" | sed -e 's/.*=> //'` | sort
----

=== Enabling and Disabling PHP Extensions

To enable or disable a PHP extension for CLI, FPM or Apache, use the commands `phpenmod` or `phpdismod`.

Example:
[source,bash]
----
sudo phpenmod php-ldap
----

or

[source,bash]
----
sudo phpdismod php-ldap
----

Type the command without parameters or see this {php-common}[guide] to read more about the commands in the php-common package.

== Notes for PHP Library phpseclib

phpseclib's BigInteger uses the php-gmp (GNU Multiple Precision Arithmetic Library), php-bcmath and OpenSSL extensions, if they're available, for speed, but doesn't require them to maintain maximum compatibility. The GMP library uses an internal resource type, while the BCMath library uses strings as datatype. The most salient difference is that GMP works on _arbitrary precision integer values_ whereas BCMath allows _arbitrary precision] decimal / float-like values_.

== Database Installation

To install a database, see the
xref:installation/manual_installation/manual_installation_db.adoc[Manual Installation Databases guide].

> NOTE: Follow the procedure described in xref:useful-tips[Useful Tips], if you want to `Disable Transparent Huge Pages (THP),Transparent Huge Pages`

Reviewer 2: ```suggestion
NOTE: Follow the procedure described in xref:useful-tips[Useful Tips] if you want to `Disable Transparent Huge Pages (THP),Transparent Huge Pages`.
```
